src/core/ext/filters/client_channel/client_channel.cc

```diff
@@ -184,12 +184,16 @@ class ChannelData {
 
   void RemoveExternalConnectivityWatcher(grpc_closure* on_complete,
                                          bool cancel) {
-    MutexLock lock(&external_watchers_mu_);
-    auto it = external_watchers_.find(on_complete);
-    if (it != external_watchers_.end()) {
-      if (cancel) it->second->Cancel();
-      external_watchers_.erase(it);
+    ExternalConnectivityWatcher* watcher = nullptr;
+    {
+      MutexLock lock(&external_watchers_mu_);
+      auto it = external_watchers_.find(on_complete);
+      if (it != external_watchers_.end()) {
+        watcher = it->second;
+        external_watchers_.erase(it);
+      }
     }
+    if (watcher != nullptr && cancel) watcher->Cancel();
```

Comment:
Please add a comment here indicating that `Cancel()` will hop into the `WorkSerializer`, so we can't be holding the mutex when we call it.

Reply:
added